File: web/docs-examples/csharp/Examples/Streaming.cs
// Copyright (c) Microsoft. All rights reserved.

using Prompty.Core;
using Prompty.OpenAI;

namespace DocsExamples.Examples;

/// <summary>
/// Streaming chat completion — consume response chunks as they arrive.
/// </summary>
public static class Streaming
{
    /// <summary>
    /// Invokes a streaming .prompty and iterates over the PromptyStream chunks.
    /// The streaming-chat.prompty sets stream:true via additionalProperties.
    /// </summary>
    public static async Task<List<object>> RunAsync(
        string promptyPath,
        Dictionary<string, object?>? inputs = null)
    {
        // Register providers
        InvokerRegistry.RegisterExecutor("openai", new OpenAIExecutor());
        InvokerRegistry.RegisterProcessor("openai", new OpenAIProcessor());

        // Load the agent — stream: true is set in the prompty's additionalProperties
        var agent = PromptyLoader.Load(promptyPath);

        // Set streaming flag in metadata (the executor checks this)
        agent.Metadata ??= new Dictionary<string, object>();
        agent.Metadata["stream"] = true;

        // Prepare messages
        var messages = await Pipeline.PrepareAsync(agent, inputs);

        // Execute returns a PromptyStream when streaming is enabled
        var response = await Pipeline.ExecuteAsync(agent, messages);

        var chunks = new List<object>();
        if (response is PromptyStream stream)
        {
            await foreach (var chunk in stream)
            {
                chunks.Add(chunk);
            }
        }

        return chunks;
    }
}

File: web/docs-examples/rust/examples/streaming.rs
use serde_json::json;

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    prompty::register_defaults();
    prompty_openai::register();

    // Load agent and prepare messages
    let agent = prompty::load("chat.prompty")?;
    let messages = prompty::prepare(&agent, Some(&json!({ "question": "Tell me a joke" }))).await?;

    // When stream: true is set, run consumes the stream and returns accumulated text
    let result = prompty::run(&agent, &messages).await?;

    println!("{result}");
    Ok(())
}
